L8-c++-internalization.md

```diff
@@ -0,0 +1,174 @@
+Separate internal-only and public sections of the C++ wrapping
+----
+* Author(s): vjpai
+* Approver: a11r
+* Status: Draft
+* Implemented in: C++
+* Last updated: June 22, 2017
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/Z8S9P4NzGHs
+
+## Abstract
+
+Maintain separation of C++ classes and functions that are intended
+to be directly used by application developers from those that are
+supposed to be used only through codegen or for that exist for gRPC's
+internal implementation (e.g., interactions with the core API).
+
+## Background
+
+Essentially all of gRPC C++ lives in `namespace grpc`, giving the
+appearance that all portions are equally intended as part of the
+publicly-exposed API. This gRFC aims to clarify this by categorizing
+the gRPC C++ wrapping, moving some components to `namespace
+grpc::internal`, and moving some class constructors to `private` (with
+friends or appropriately-named `static` factories for internalized generation).
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+The C++ gRPC wrapping includes numerous classes and functions, but
+they fall into four separate categories:
+
+1. Documented for public use
+1. Intended for interfacing with serialization layers such as protobuf
+1. Intended for use through the code-generation layer
+1. Intended for use in the internal implementation of gRPC C++
+1. Intended for interfacing with gRPC core
+
+This gRFC aims to clarify that the first two are intended for public
+use (the second one to support custom serializers), but that the last
+three should not be used as such. This will be achived by
+moving classes and functions in the latter three categories to `namespace
+grpc::internal`.
+
+* Accessed through codegen 
+  - `RpcMethod`
+  - `BlockingUnaryCall`
+
+* gRPC C++ implementation
+  - `CallHook`
+  - `CallOpSetInterface`
+  - `CallOpSet`
+  - `SneakyCallOpSet`
+  - `CallNoOp`
+  - `RpcMethodHandler`
+  - `UnknownMethodHandler`
+  - `ClientStreamingHandler`
+  - `ServerStreamingHandler`
+  - `TemplatedBidiStreamingHandler`
+  - `BidiStreamingHandler`
+  - `StreamedUnaryHandler`
+  - `SplitServerStreamingHandler`
+
+* gRPC C++ interface with core
+  - `CompletionQueueTag`
+  - `Call`
+  - `CallOpSendInitialMetadata`
+  - `CallOpSendMessage`
+  - `CallOpClientSendClose`
+  - `CallOpServerSendStatus`
+  - `CallOpClientRecvStatus`
+  - `CallOpRecvInitialMetadata`
+  - `CallOpRecvMessage`
+  - `CallOpGenericRecvMessage`
+  - `MetadataMap`
+
+Additionally, some classes in the first category are only supposed to
+be created through the code generator or method handlers. To support
+this, we will privatize their constructors and allow them to be
+created only by internalized `friend` classes or through `static`
+factories invoked by the code generator (which are themselves moved to
+a `struct internal`).
```

Comment:
I should have clarified: `struct internal` is defined separately in each of the classes that is being modified. So instead of calling `grpc::ClientAsyncResponseReader::Create`, you'll call `grpc::ClientAsyncResponseReader::internal::Create` . It's not one big `struct internal` which the document might have made it look like.